server/src/main/java/org/apache/lucene/index/LazySoftDeletesDirectoryReaderWrapper.java

```diff
@@ -0,0 +1,247 @@
+        final int numDocs = maxDoc - segmentInfo.getDelCount() - segmentInfo.getSoftDelCount();
+        final Bits lazyBits = new Bits() {
+
+            volatile Bits materializedBits = null;
+
+            @Override
+            public boolean get(int index) {
+                if (materializedBits == null) {
+                    synchronized (this) {
+                        try {
+                            if (materializedBits == null) {
+                                materializedBits = init();
+                            }
+                        } catch (IOException e) {
+                            throw new UncheckedIOException(e);
+                        }
+                    }
+                }
+                return materializedBits.get(index);
+            }
+
+            @Override
+            public int length() {
+                return maxDoc;
+            }
+
+            private Bits init() throws IOException {
+                assert Thread.holdsLock(this);
+
+                DocIdSetIterator iterator = DocValuesFieldExistsQuery.getDocValuesDocIdSetIterator(field, reader);
+                assert iterator != null;
+                Bits liveDocs = reader.getLiveDocs();
+                final FixedBitSet bits;
+                if (liveDocs != null) {
+                    bits = FixedBitSet.copyOf(liveDocs);
+                } else {
+                    bits = new FixedBitSet(reader.maxDoc());
+                    bits.set(0, reader.maxDoc());
+                }
+                int numSoftDeletes = PendingSoftDeletes.applySoftDeletes(iterator, bits);
+                int numDeletes = reader.numDeletedDocs() + numSoftDeletes;
+                int numDocs = reader.maxDoc() - numDeletes;
+                assert assertDocCounts(numDocs, numSoftDeletes, reader);
+                return bits;
+            }
+        };
+        return new LazySoftDeletesFilterCodecReader(segmentReader, lazyBits, numDocs);
+    }
+
+    private static boolean assertDocCounts(int expectedNumDocs, int numSoftDeletes, LeafReader reader) {
+        if (reader instanceof SegmentReader) {
+            SegmentReader segmentReader = (SegmentReader) reader;
+            SegmentCommitInfo segmentInfo = segmentReader.getSegmentInfo();
+            assert segmentReader.isNRT == false;
+            int numDocs = segmentInfo.info.maxDoc() - segmentInfo.getSoftDelCount() - segmentInfo.getDelCount();
+            assert numDocs == expectedNumDocs : "numDocs: " + numDocs + " expected: " + expectedNumDocs
+                + " maxDoc: " + segmentInfo.info.maxDoc()
+                + " getDelCount: " + segmentInfo.getDelCount()
+                + " getSoftDelCount: " + segmentInfo.getSoftDelCount()
+                + " numSoftDeletes: " + numSoftDeletes
+                + " reader.numDeletedDocs(): " + reader.numDeletedDocs();
+            // in the NRT case we don't have accurate numbers for getDelCount and getSoftDelCount since they might not be
```

Comment:
I think this class is used to wrap only non-NRT readers (assertion in line 173)?  Hence, it's safe to assert the numDocs here. For NRT readers, IndexWriter atomically handles soft-deletes.